File: Manipulator3D_Julia/src/sim/Trajectory.jl
module Sim

export LinearTrajectory, reset!, update!, position, finished, alpha

using ..Robot: Vec3, v3

mutable struct LinearTrajectory
    a::Vec3
    b::Vec3
    duration::Float32
    t::Float32
    α::Float32
    done::Bool
end

function LinearTrajectory()
    return LinearTrajectory(v3(0,0,0), v3(0,0,0), 1.0f0, 0.0f0, 0.0f0, true)
end

function reset!(tr::LinearTrajectory, from::Vec3, to::Vec3, durationSec::Real)
    tr.a = from
    tr.b = to
    d = Float32(durationSec)
    tr.duration = (d > 1f-6) ? d : 1f-6
    tr.t = 0.0f0
    tr.α = 0.0f0
    tr.done = false
    return nothing
end

function update!(tr::LinearTrajectory, dt::Float32)
    tr.done && return nothing
    tr.t += dt
    tr.α = clamp(tr.t / tr.duration, 0.0f0, 1.0f0)
    if tr.α >= 1.0f0
        tr.done = true
    end
    return nothing
end

function position(tr::LinearTrajectory)::Vec3
    a = tr.a
    b = tr.b
    α = tr.α
    return v3(
        a.x + (b.x - a.x) * α,
        a.y + (b.y - a.y) * α,
        a.z + (b.z - a.z) * α
    )
end

finished(tr::LinearTrajectory) = tr.done
alpha(tr::LinearTrajectory) = tr.α

end # module Sim

File: Manipulator3D_Go/src/sim/trajectory.go
package sim

import "manipulator3d/src/robot"

type LinearTrajectory struct {
	a        robot.Vec3
	b        robot.Vec3
	duration float32
	t        float32
	alpha    float32
	finished bool
}

func NewLinearTrajectory() *LinearTrajectory {
	return &LinearTrajectory{
		a:        robot.Vec3{},
		b:        robot.Vec3{},
		duration: 1.0,
		t:        0,
		alpha:    0,
		finished: true,
	}
}

func (tr *LinearTrajectory) Reset(from, to robot.Vec3, durationSec float32) {
	tr.a = from
	tr.b = to

	if durationSec < 1e-6 {
		durationSec = 1e-6
	}
	tr.duration = durationSec
	tr.t = 0
	tr.alpha = 0
	tr.finished = false
}

func (tr *LinearTrajectory) Update(dt float32) {
	if tr.finished {
		return
	}
	tr.t += dt
	tr.alpha = clamp01(tr.t / tr.duration)
	if tr.alpha >= 1.0 {
		tr.finished = true
	}
}

func (tr *LinearTrajectory) Position() robot.Vec3 {
	return robot.Vec3{
		X: tr.a.X + (tr.b.X-tr.a.X)*tr.alpha,
		Y: tr.a.Y + (tr.b.Y-tr.a.Y)*tr.alpha,
		Z: tr.a.Z + (tr.b.Z-tr.a.Z)*tr.alpha,
	}
}

func (tr *LinearTrajectory) Finished() bool { return tr.finished }
func (tr *LinearTrajectory) Alpha() float32 { return tr.alpha }

func clamp01(x float32) float32 {
	if x < 0 {
		return 0
	}
	if x > 1 {
		return 1
	}
	return x
}
